swarm/src/behaviour.rs

```diff
@@ -721,3 +849,264 @@ impl Default for CloseConnection {
         CloseConnection::All
     }
 }
+
+/// Enumeration with the list of the possible events
+/// to pass to [`on_swarm_event`](NetworkBehaviour::on_swarm_event).
+pub enum FromSwarm<'a, Handler: IntoConnectionHandler> {
+    /// Informs the behaviour about a newly established connection to a peer.
+    ConnectionEstablished(ConnectionEstablished<'a>),
+    /// Informs the behaviour about a closed connection to a peer.
+    ///
+    /// This event is always paired with an earlier
+    /// [`FromSwarm::ConnectionEstablished`] with the same peer ID, connection ID
+    /// and endpoint.
+    ConnectionClosed(ConnectionClosed<'a, Handler>),
+    /// Informs the behaviour that the [`ConnectedPoint`] of an existing
+    /// connection has changed.
+    AddressChange(AddressChange<'a>),
+    /// Informs the behaviour that the dial to a known
+    /// or unknown node failed.
+    DialFailure(DialFailure<'a, Handler>),
+    /// Informs the behaviour that an error
+    /// happened on an incoming connection during its initial handshake.
+    ///
+    /// This can include, for example, an error during the handshake of the encryption layer, or the
+    /// connection unexpectedly closed.
+    ListenFailure(ListenFailure<'a, Handler>),
+    /// Informs the behaviour that a new listener was created.
+    NewListener(NewListener),
+    /// Informs the behaviour that we have started listening on a new multiaddr.
+    NewListenAddr(NewListenAddr<'a>),
+    /// Informs the behaviour that a multiaddr
+    /// we were listening on has expired,
+    /// which means that we are no longer listening on it.
+    ExpiredListenAddr(ExpiredListenAddr<'a>),
+    /// Informs the behaviour that a listener experienced an error.
+    ListenerError(ListenerError<'a>),
+    /// Informs the behaviour that a listener closed.
+    ListenerClosed(ListenerClosed<'a>),
+    /// Informs the behaviour that we have discovered a new external address for us.
+    NewExternalAddr(NewExternalAddr<'a>),
+    /// Informs the behaviour that an external address was removed.
+    ExpiredExternalAddr(ExpiredExternalAddr<'a>),
+}
+
+/// [`FromSwarm`] variant that informs the behaviour about a newly established connection to a peer.
+#[derive(Clone, Copy)]
+pub struct ConnectionEstablished<'a> {
+    pub peer_id: PeerId,
+    pub connection_id: ConnectionId,
+    pub endpoint: &'a ConnectedPoint,
+    pub failed_addresses: &'a [Multiaddr],
+    pub other_established: usize,
+}
+
+/// [`FromSwarm`] variant that informs the behaviour about a closed connection to a peer.
+///
+/// This event is always paired with an earlier
+/// [`FromSwarm::ConnectionEstablished`] with the same peer ID, connection ID
+/// and endpoint.
+pub struct ConnectionClosed<'a, Handler: IntoConnectionHandler> {
+    pub peer_id: PeerId,
+    pub connection_id: ConnectionId,
+    pub endpoint: &'a ConnectedPoint,
+    pub handler: <Handler as IntoConnectionHandler>::Handler,
```

Comment:
The variants which have the `Handler` also prevent our enum from being `cloned`. Aren't these semantically more an event that is coming from the connection handler and should thus be merged with the `on_connection_handler_event` callback?

Reply:
Off the top of my head, I would see these events to better fit `on_swarm_event`. An event coming from the `ConnectionHandler` containing a `ConnectionHandler` seems off to me.

Reply:
> Off the top of my head, I would see these events to better fit `on_swarm_event`. An event coming from the `ConnectionHandler` containing a `ConnectionHandler` seems off to me.

What if we have a third callback? If we look at the swarm-derive closely, we can see that there is a difference here.

Some events are given as is to all behaviours. For events that contain a connection handler, each behaviour only receives the handler it originally created.

If we split these into different callbacks (and assuming we can get rid of the listener error), then the non-handler events could just be cloned which would make forwarding much easier!

Reply:
> Some events are given as is to all behaviours. For events that contain a connection handler, each behaviour only receives the handler it originally created.

Agreed that there is a difference here.

> What if we have a third callback? If we look at the swarm-derive closely, we can see that there is a difference here.

I would need to give this more thought. Off the top of my head the above does not justify an additional hook.

I would prefer this being explored in a separate pull request. Would that be OK for you @thomaseizinger?

Reply:
> > What if we have a third callback? If we look at the swarm-derive closely, we can see that there is a difference here.
> 
> I would need to give this more thought. Off the top of my head the above does not justify an additional hook.
> 
> I would prefer this being explored in a separate pull request. Would that be OK for you @thomaseizinger?

Yeah we can explore that on top of this for sure.